### PR TITLE
LibJS: Implement the encode/decodeURI(Component) family of functions 

### PR DESCRIPTION
**Pull request by IdanHo**
These are generally useful and in particular needed for twitter.com

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/GlobalObject.cpp [hunk 1]
```diff
@@ -340,4 +346,117 @@ JS_DEFINE_NATIVE_FUNCTION(GlobalObject::eval)
...
+        }
+        decoded_builder.append(decoded_code_unit);
+        expected_continuation_bytes = leading_ones - 1;
+    }
+    return decoded_builder.build();
+}
+
+JS_DEFINE_NATIVE_FUNCTION(GlobalObject::encode_uri)
+{
+    auto uri_string = (vm.argument_count() > 0 ? vm.argument(0) : js_undefined()).to_string(global_object);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/GlobalObject.cpp:420]: `vm.argument()` returns undefined on OOB access, so we can simplify this:

```suggestion
    auto uri_string = vm.argument(0).to_string(global_object);
```

File: Userland/Libraries/LibJS/Runtime/GlobalObject.cpp [hunk 2]
```diff
@@ -340,4 +346,117 @@ JS_DEFINE_NATIVE_FUNCTION(GlobalObject::eval)
...
+        return {};
+    auto encoded = encode(global_object, uri_string, ";/?:@&=+$,abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_.!~*'()#"sv);
+    if (vm.exception())
+        return {};
+    return js_string(vm, move(encoded));
+}
+
+JS_DEFINE_NATIVE_FUNCTION(GlobalObject::decode_uri)
+{
+    auto uri_string = (vm.argument_count() > 0 ? vm.argument(0) : js_undefined()).to_string(global_object);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/GlobalObject.cpp:431]:
```suggestion
    auto uri_string = vm.argument(0).to_string(global_object);
```

File: Userland/Libraries/LibJS/Runtime/GlobalObject.cpp [hunk 3]
```diff
@@ -340,4 +346,117 @@ JS_DEFINE_NATIVE_FUNCTION(GlobalObject::eval)
...
+        return {};
+    auto decoded = decode(global_object, uri_string, ";/?:@&=+$,#"sv);
+    if (vm.exception())
+        return {};
+    return js_string(vm, move(decoded));
+}
+
+JS_DEFINE_NATIVE_FUNCTION(GlobalObject::encode_uri_component)
+{
+    auto uri_string = (vm.argument_count() > 0 ? vm.argument(0) : js_undefined()).to_string(global_object);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/GlobalObject.cpp:442]:
```suggestion
    auto uri_string = vm.argument(0).to_string(global_object);
```

File: Userland/Libraries/LibJS/Runtime/GlobalObject.cpp [hunk 4]
```diff
@@ -340,4 +346,117 @@ JS_DEFINE_NATIVE_FUNCTION(GlobalObject::eval)
...
+        return {};
+    auto encoded = encode(global_object, uri_string, "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_.!~*'()"sv);
+    if (vm.exception())
+        return {};
+    return js_string(vm, move(encoded));
+}
+
+JS_DEFINE_NATIVE_FUNCTION(GlobalObject::decode_uri_component)
+{
+    auto uri_string = (vm.argument_count() > 0 ? vm.argument(0) : js_undefined()).to_string(global_object);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/GlobalObject.cpp:453]:
```suggestion
    auto uri_string = vm.argument(0).to_string(global_object);
```

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibJS/Runtime/GlobalObject.cpp (hunk 1 above)
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/GlobalObject.cpp]: fixed, thanks!

**Pull request merged**